forge/src/main/java/me/lucko/luckperms/forge/ForgeSenderFactory.java

```diff
@@ -0,0 +1,110 @@
+ *  furnished to do so, subject to the following conditions:
+ *
+ *  The above copyright notice and this permission notice shall be included in all
+ *  copies or substantial portions of the Software.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ *  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ *  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ *  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ *  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ *  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ *  SOFTWARE.
+ */
+
+package me.lucko.luckperms.forge;
+
+import me.lucko.luckperms.common.cacheddata.result.TristateResult;
+import me.lucko.luckperms.common.locale.TranslationManager;
+import me.lucko.luckperms.common.query.QueryOptionsImpl;
+import me.lucko.luckperms.common.sender.Sender;
+import me.lucko.luckperms.common.sender.SenderFactory;
+import me.lucko.luckperms.common.verbose.VerboseCheckTarget;
+import me.lucko.luckperms.common.verbose.event.CheckOrigin;
+import me.lucko.luckperms.forge.bridge.server.level.ServerPlayerBridge;
+import net.kyori.adventure.text.Component;
+import net.kyori.adventure.text.serializer.gson.GsonComponentSerializer;
+import net.luckperms.api.util.Tristate;
+import net.minecraft.commands.CommandSourceStack;
+import net.minecraft.server.level.ServerPlayer;
+
+import java.util.Locale;
+import java.util.UUID;
+
+public class ForgeSenderFactory extends SenderFactory<LPForgePlugin, CommandSourceStack> {
+    public ForgeSenderFactory(LPForgePlugin plugin) {
+        super(plugin);
+    }
+
+    @Override
+    protected UUID getUniqueId(CommandSourceStack commandSource) {
+        if (commandSource.getEntity() != null) {
+            return commandSource.getEntity().getUUID();
+        }
+        return Sender.CONSOLE_UUID;
+    }
+
+    @Override
+    protected String getName(CommandSourceStack commandSource) {
+        String name = commandSource.getTextName();
+        if (commandSource.getEntity() != null && name.equals("Server")) {
+            return Sender.CONSOLE_NAME;
+        }
+        return name;
+    }
+
+    @Override
+    protected void sendMessage(CommandSourceStack sender, Component message) {
+        Locale locale;
+        if (sender.getEntity() instanceof ServerPlayer) {
+            locale = ((ServerPlayerBridge) sender.getEntity()).bridge$getLocale();
+        } else {
+            locale = null;
+        }
+
+        sender.sendSuccess(toNativeText(TranslationManager.render(message, locale)), false);
+    }
+
+    @Override
+    protected Tristate getPermissionValue(CommandSourceStack commandSource, String node) {
+        if (commandSource.getEntity() instanceof ServerPlayer) {
+            return ((ServerPlayerBridge) commandSource.getEntity()).bridge$hasPermission(node);
```

Comment:
Can we use the Forge permissions API here instead?

Would be good to minimize Mixin use, as they are the main thing that cause problems on game/Forge version updates.

Reply:
The Forge [PermissionAPI.getPermission](https://github.com/MinecraftForge/MinecraftForge/blob/335cb5edc4467a213a01a315e75aad5aa3569f34/src/main/java/net/minecraftforge/server/permission/PermissionAPI.java#L67-L81) requires that the node being looked up has been registered. Not sure if this is doable